File: plugins/plugin-nextcloud-talk/typescript/src/providers/chatState.ts
import type { IAgentRuntime, Memory, Provider, ProviderResult, State } from "@elizaos/core";
import { NEXTCLOUD_TALK_SERVICE_NAME } from "../constants";
import type { NextcloudTalkService } from "../service";

export const CHAT_STATE_PROVIDER = "nextcloud_talk_chat_state";

export const chatStateProvider: Provider = {
  name: CHAT_STATE_PROVIDER,
  description:
    "Provides Nextcloud Talk chat context including room token, sender ID, and room type",
  dynamic: true,

  get: async (runtime: IAgentRuntime, message: Memory, _state: State): Promise<ProviderResult> => {
    const roomToken = message.content?.roomToken as string | undefined;
    const senderId = message.content?.senderId as string | undefined;
    const roomId = message.roomId;

    const service = runtime.getService(NEXTCLOUD_TALK_SERVICE_NAME) as
      | NextcloudTalkService
      | undefined;
    const room = roomToken ? service?.getRoom(roomToken) : undefined;

    const isGroupChat = room?.type === "group" || room?.type === "public";
    const isPrivate = room?.type === "one-to-one";

    const data = {
      roomToken,
      senderId,
      roomId,
      roomName: room?.displayName || room?.name,
      roomType: room?.type,
      isGroupChat,
      isPrivate,
      baseUrl: service?.baseUrl,
    };

    const values: Record<string, string> = {
      room_token: roomToken || "",
      sender_id: senderId || "",
      room_id: roomId || "",
      room_name: room?.displayName || room?.name || "",
      room_type: room?.type || "",
      is_group_chat: isGroupChat.toString(),
      is_private: isPrivate.toString(),
      base_url: service?.baseUrl || "",
    };

    let text = "Nextcloud Talk Chat State:\n";
    if (roomToken) {
      text += `Room Token: ${roomToken}\n`;
      text += `Room Name: ${room?.displayName || room?.name || "Unknown"}\n`;
      text += `Room Type: ${isPrivate ? "Private (1:1)" : isGroupChat ? "Group" : "Unknown"}\n`;
    }
    if (senderId) {
      text += `Sender ID: ${senderId}\n`;
    }
    if (service?.baseUrl) {
      text += `Nextcloud URL: ${service.baseUrl}\n`;
    }

    return {
      data,
      values,
      text,
    };
  },
};

File: plugins/plugin-nextcloud-talk/rust/src/providers/chat_state.rs
use async_trait::async_trait;
use serde_json::Value;

use super::{NextcloudTalkProvider, ProviderContext};

/// Provider that exposes the current chat/room/user context as JSON.
pub struct ChatStateProvider;

#[async_trait]
impl NextcloudTalkProvider for ChatStateProvider {
    fn name(&self) -> &'static str {
        "nextcloud_talk_chat_state"
    }

    async fn get(&self, context: &ProviderContext) -> Value {
        let is_group = context.is_group_chat.unwrap_or(false);
        let is_private = context.is_group_chat.map(|g| !g).unwrap_or(false);

        serde_json::json!({
            "room_token": context.room_token,
            "user_id": context.user_id,
            "room_name": context.room_name,
            "room_id": context.room_id,
            "is_group_chat": is_group,
            "is_private": is_private
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_chat_state_group() {
        let provider = ChatStateProvider;
        let context = ProviderContext {
            room_token: Some("abc123".to_string()),
            user_id: Some("user1".to_string()),
            room_name: Some("Test Group".to_string()),
            room_id: Some("room-uuid".to_string()),
            is_group_chat: Some(true),
        };

        let state = provider.get(&context).await;
        assert_eq!(state["room_token"], "abc123");
        assert_eq!(state["is_group_chat"], true);
        assert_eq!(state["is_private"], false);
    }

    #[tokio::test]
    async fn test_chat_state_private() {
        let provider = ChatStateProvider;
        let context = ProviderContext {
            room_token: Some("xyz789".to_string()),
            user_id: Some("user2".to_string()),
            room_name: Some("John Doe".to_string()),
            room_id: Some("room-uuid".to_string()),
            is_group_chat: Some(false),
        };

        let state = provider.get(&context).await;
        assert_eq!(state["room_token"], "xyz789");
        assert_eq!(state["is_group_chat"], false);
        assert_eq!(state["is_private"], true);
    }
}
